libraries/AP_FETtecOneWire/AP_FETtecOneWire.cpp

```diff
@@ -0,0 +1,831 @@
+{
+    // only accept messages from the bootloader when we could
+    // legitimately get a message from the bootloader.  Swipes the OK
+    // message for convenience
+    const FrameSource frame_source = (FrameSource)u.packed_ok.frame_source;
+    if (frame_source != FrameSource::ESC) {
+        if (esc.state != ESCState::WAITING_OK_FOR_RUNNING_SW_TYPE) {
+            return;
+        }
+    }
+
+    switch (esc.state) {
+    case ESCState::WANT_SEND_OK_TO_GET_RUNNING_SW_TYPE:
+        return;
+    case ESCState::WAITING_OK_FOR_RUNNING_SW_TYPE:
+        // "OK" is the only valid response
+        if (!buffer_contains_ok(length)) {
+            return;
+        }
+        switch (frame_source) {
+        case FrameSource::MASTER:
+            // probably half-duplex; should be caught before we get here
+            INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
+            break;
+        case FrameSource::BOOTLOADER:
+            esc.set_state(ESCState::WANT_SEND_START_FW);
+            break;
+        case FrameSource::ESC:
+#if HAL_AP_FETTEC_ONEWIRE_GET_STATIC_INFO
+            esc.set_state(ESCState::WANT_SEND_REQ_TYPE);
+#else
+#if HAL_WITH_ESC_TELEM
+            esc.set_state(ESCState::WANT_SEND_SET_TLM_TYPE);
+#else
+            esc.set_state(ESCState::WANT_SEND_SET_FAST_COM_LENGTH);
+#endif
+#endif
+            break;
+        }
+        break;
+
+    case ESCState::WANT_SEND_START_FW:
+        return;
+    case ESCState::WAITING_OK_FOR_START_FW:
+        if (buffer_contains_ok(length)) {
+#if HAL_AP_FETTEC_ONEWIRE_GET_STATIC_INFO
+            esc.set_state(ESCState::WANT_SEND_REQ_TYPE);
+#else
+#if HAL_WITH_ESC_TELEM
+            esc.set_state(ESCState::WANT_SEND_SET_TLM_TYPE);
+#else
+            esc.set_state(ESCState::WANT_SEND_SET_FAST_COM_LENGTH);
+#endif
+#endif
+        }
+        break;
+
+#if HAL_AP_FETTEC_ONEWIRE_GET_STATIC_INFO
+    case ESCState::WANT_SEND_REQ_TYPE:
+        return;
+    case ESCState::WAITING_ESC_TYPE:
+        if (length != sizeof(u.packed_esc_type)) {
+            _message_invalid_in_state_count++;
+            return;
+        }
+        esc.type = u.packed_esc_type.msg.type;
+        esc.set_state(ESCState::WANT_SEND_REQ_SW_VER);
+        break;
+
+    case ESCState::WANT_SEND_REQ_SW_VER:
+        return;
+    case ESCState::WAITING_SW_VER:
+        if (length != sizeof(u.packed_sw_ver)) {
+            _message_invalid_in_state_count++;
+            return;
+        }
+        esc.firmware_version = u.packed_sw_ver.msg.version;
+        esc.firmware_subversion = u.packed_sw_ver.msg.subversion;
+        esc.set_state(ESCState::WANT_SEND_REQ_SN);
+        break;
+
+    case ESCState::WANT_SEND_REQ_SN:
+        return;
+    case ESCState::WAITING_SN:
+        if (length != sizeof(u.packed_sn)) {
+            _message_invalid_in_state_count++;
+            return;
+        }
+        memset(esc.serial_number, '\0', ARRAY_SIZE(esc.serial_number));

```

Comment:
Why all this? why not just copy the 12 bytes? why the MIN() call in the constructor?

Reply:
We do not want to either over-write or over-read in the case that the array sizes differ.  They really, really shouldn't - but there's no runtime cost to sizing the memcpy appropriately.   But it does mean you have to zero the destination in case the source is shorter than the dest.

This could also be done with a static assert that the arrays are the same size.